homeassistant/components/mysensors/device.py

```diff
@@ -19,33 +32,99 @@
 MYSENSORS_PLATFORM_DEVICES = "mysensors_devices_{}"
 
 
-def get_mysensors_devices(hass, domain):
-    """Return MySensors devices for a platform."""
-    if MYSENSORS_PLATFORM_DEVICES.format(domain) not in hass.data:
-        hass.data[MYSENSORS_PLATFORM_DEVICES.format(domain)] = {}
-    return hass.data[MYSENSORS_PLATFORM_DEVICES.format(domain)]
-
-
 class MySensorsDevice:
     """Representation of a MySensors device."""
 
-    def __init__(self, gateway, node_id, child_id, name, value_type):
+    def __init__(
+        self,
+        gateway_id: GatewayId,
+        gateway: BaseAsyncGateway,
+        node_id: int,
+        child_id: int,
+        value_type: int,
+    ):
         """Set up the MySensors device."""
-        self.gateway = gateway
-        self.node_id = node_id
-        self.child_id = child_id
-        self._name = name
-        self.value_type = value_type
-        child = gateway.sensors[node_id].children[child_id]
-        self.child_type = child.type
+        self.gateway_id: GatewayId = gateway_id
+        self.gateway: BaseAsyncGateway = gateway
+        self.node_id: int = node_id
+        self.child_id: int = child_id
+        self.value_type: int = value_type  # value_type as int. string variant can be looked up in gateway consts
+        self.child_type = self._mysensors_childsensor.type
         self._values = {}
         self._update_scheduled = False
         self.hass = None
 
+    @property
+    def dev_id(self) -> DevId:
+        """Return the DevId of this device.
+
+        It is used to route incoming MySensors messages to the correct device/entity.
+        """
+        return self.gateway_id, self.node_id, self.child_id, self.value_type
+
+    @property
+    def _logger(self):
+        return logging.getLogger(f"{__name__}.{self.name}")
+
+    async def async_will_remove_from_hass(self):
+        """Remove this entity from home assistant."""
+        for platform in PLATFORM_TYPES:
+            platform_str = MYSENSORS_PLATFORM_DEVICES.format(platform)
+            if platform_str in self.hass.data[DOMAIN]:
+                platform_dict = self.hass.data[DOMAIN][platform_str]
+                if self.dev_id in platform_dict:
+                    if platform_dict[self.dev_id] is not self:
```

Comment:
When can this happen?

Reply:
It shouldn't. If it did, that would be a bug.

A bug that triggers this would mean a serious fault in the DevId -> MySensorsDevice mapping, which would probably manifest in lots of weird behaviour because messages would be delivered to the wrong entities. The check is quite cheap to run and would help immensely in such a situation.

I would recommend to keep it in, but of course if you don't like it i will remove it.

Reply:
Bump.